book/src/ephemeral-local-volumes.md

```diff
@@ -1,38 +1,49 @@
-# Ephemeral Local Volumes
+# Pod Inline Volume Support
 
-> ## *This page is still under active development.*
+## Status
 
-**Status:** Alpha
+Status | Min K8s Version | Max K8s Version 
+--|--|--
+Alpha | 1.15 | - 
 
-Kubernetes supports three types of volumes:
+## Overview
+Traditionally, volumes that are backed by CSI drivers can only be used with a `PersistentVolume` and `PersistentVolumeClaim` object combination. This feature supports ephemeral storage use cases and allows CSI volumes to be specified directly in the pod specification.  At runtime, nested inline volumes follow the ephemeral lifecycle of their associated pods where the driver handles all phases of volume operations as pods are created and destroyed.
 
-1. Remote Persistent Volumes
-2. Local Persistent Volumes
-3. Local Ephemeral Volumes
+> See the [design document](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md) for futher information.
 
-The initial focus of Kubernetes CSI was Remote Persistent Volumes. However, the goal is for CSI to support all three types.
 
-This page documents how to create "Local Ephemeral Volumes" for Kubernetes using CSI.
+### Example of inline CSI pod spec
+A pod spec with an ephemeral inline CSI volume. Note that because the volume is expected to be ephemeral, the volumeHandle is not provided. Instead a CSI-generated ID will be submitted to the driver.
 
-# What is a Local Ephemeral Volumes?
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: some-pod
+spec:
+  containers:
+    ...
+  volumes:
+      - name: vol
+        csi:
+          driver: inline.storage.kubernetes.io
+          volumeAttributes:
+              foo: bar
+```
 
-A Local Ephemeral Volumes is a volume whose lifecycle is tied to the lifecycle of a single pod:
+## Implementing inline ephemeral support
 
-* The volume is "provisioned" (either empty or with some pre-populated data) when the pod is created.
-* The volume is deleted when the pod is terminated.
+Drivers must be modified (or implemented specifically) to support inline ephmeral workflows.  When Kubernetes encounters an inline CSI volume embedded in a pod spec, it treats that volume differently.  Mainly, the driver will only receive `NodePublish`, during the volume's mount phase, and `NodeUnpublish` when the pod is going away and the volume is unmounted.   To support inline, a driver must implement the followings:
 
-[Kubernetes Secret Volumes](https://kubernetes.io/docs/concepts/storage/volumes/#secret) are a good example (non-CSI) of a local ephemeral volumes.
+- Identity service
+- Node service
```

Comment:
As far as I understand it, there's more that is needed during deployment to make this work:
- configure `PodSecurityPolicy` (https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md#specifying-allowed-inline-drivers-with-podsecuritypolicy)
- define driver mode (https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md#driver-mode)

It would be useful to mention this here.

Note that the KEP is inconsistent regarding `PodSecurityPolicy`: the comment says that "Empty or nil indicates that all CSI drivers may be used." but the plain text then says "An empty value means no CSI drivers are allowed to be specified inline inside a pod spec." - which one is correct?

Reply:
PodSecurityPolicy is only required if you actually have a PodSecurityPolicy.

Driver mode is in the spec but not currently implemented. In the alpha, it just assumes it supports ephemeral if you try and use it in pod inline mode.

Reply:
We did not document driver mode here because it's currently unimplemented.

Reply:
Thanks for helping clarify this.

Reply:
i am trying to implement inline volume support for cinder csi driver, few queries here since mode is not supported and  `feature-gates=CSIInlineVolume=true` , is it assumed a driver is supporting both modes persistent+ephemeral??

Reply:
One more thing i got from inline KEP, if requested volume is ephemeral, it will be notified in `NodePublishRequest.volume_context ` as `csi.storage.k8s.io/ephemeral: true`, it would be useful if this information is added in the doc for driver developers looking to support both persistent+ephemeral

Reply:
ok, got it from here https://github.com/kubernetes-csi/docs/pull/170/files 
`csi.storage.k8s.io/ephemeral: true`  is not supported in v1.15, **so as of v1.15 at one time driver can be run in only one mode persistent or ephemeral,** could we specify this information in the doc?

Reply:
> i am trying to implement inline volume support for cinder csi driver, few queries here since mode is not supported and `feature-gates=CSIInlineVolume=true` , is it assumed a driver is supporting both modes persistent+ephemeral??

That feature to support both is being worked on by @pohly for 1.16. Current behavior is to support one or the other.

Reply:
> One more thing i got from inline KEP, if requested volume is ephemeral, it will be notified in `NodePublishRequest.volume_context ` as `csi.storage.k8s.io/ephemeral: true`, it would be useful if this information is added in the doc for driver developers looking to support both persistent+ephemeral

Thanks.   That is a good point, please open an issue against the github.com/kubernetes-csi/docs so this can be tracked.

Reply:
For now let's only document what's possible as of 1.15.  we'll update this again once 1.16 is released

Reply:
@msau42 agreed, but IMO as of 1.15 it should be explicitly specified in doc `that a driver can be support one mode only, feature to support both modes in parallel will be added in v1.16,`
 i have seen it is specified in the example section below of hostpath driver, but from there it is inferred as hostpath is developed in such a way, it is the limitation of the driver.

Reply:
I'll get to that soon. Let's merge this revision of the docs as-is.